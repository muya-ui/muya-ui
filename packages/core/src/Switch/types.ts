import React from 'react';

import { ICustomStyleBaseProps, ISizeSpecBaseProps, Omit } from '../types';

/**
 * Switch style keys
 * @styles loadingWrapper 加载的容器
 * @styles childrenWrapper 子节点容器
 */
export type ISwitchStyleKeys = 'loadingWrapper' | 'childrenWrapper';

export interface ISwitchBaseProps
  extends ISizeSpecBaseProps,
    ICustomStyleBaseProps<ISwitchStyleKeys> {
  /**
   * 禁用状态
   * @default false
   */
  disabled?: boolean;
  /**
   * 选中的子节点
   */
  checkedChildren?: React.ReactNode;
  /**
   * 未选中的子节点
   */
  unCheckedChildren?: React.ReactNode;
  /**
   * change 事件
   */
  onChange?: (checked: boolean, e: React.MouseEvent | React.KeyboardEvent) => void;
  /**
   * mouseUp 事件
   */
  onMouseUp?: React.MouseEventHandler;
  /**
   * click 事件
   */
  onClick?: (checked: boolean, e: React.MouseEvent) => void;
  /**
   * 选中状态
   */
  checked?: boolean;
  /**
   * 默认选中状态
   * @default false
   */
  defaultChecked?: boolean;
  /**
   * 是否自动获取焦点
   * @default false
   */
  autoFocus?: boolean;
  /**
   * 加载 Icon
   */
  loadingIcon?: React.ReactNode;
  /**
   * 加载状态
   */
  loading?: boolean;
}

export type ISwitchProps = ISwitchBaseProps &
  Omit<React.HTMLAttributes<HTMLDivElement>, 'onChange'>;
