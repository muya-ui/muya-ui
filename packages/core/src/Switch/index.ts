export { default } from './Switch';

export * from './types';
export { default as Switch } from './Switch';
export * from './Switch';
