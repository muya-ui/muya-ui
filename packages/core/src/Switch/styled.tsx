import styled, { css } from 'styled-components';

import { IThemedBaseProps } from '../types';
import { ISwitchProps } from './types';

export interface IStyledSwitchProps
  extends Pick<ISwitchProps, 'disabled' | 'checked' | 'loading' | 'size'>,
    IThemedBaseProps {}

export const StyledSwitchChildrenWrapper = styled.div`
  display: block;
`;

export const StyledLoadingIconWrapper = styled.span`
  z-index: 1;
  font-size: 12px;
  background: transparent;
`;

export const switchCss = (props: IStyledSwitchProps) => {
  const {
    size,
    theme: {
      colors: { pattern: colorsPattern },
      opacity: { pattern: opacityPattern },
      transition: {
        pattern: { duration, easing },
      },
      components: { Switch: token },
    },
    checked,
    disabled,
  } = props;
  const iconFontSize = token.height[size!] - 12;
  return css`
    position: relative;
    display: inline-block;
    box-sizing: border-box;
    height: ${token.height[size!]}px;
    line-height: ${token.height[size!] - 4}px;
    min-width: ${token.minWidth[size!]}px;
    background-color: ${checked ? token.background.checked : token.background.unChecked};
    border-radius: 999rem;
    vertical-align: middle;
    border: 2px solid transparent;
    opacity: ${disabled ? opacityPattern.disabled : 1};
    cursor: ${disabled ? 'not-allowed' : 'pointer'};
    user-select: none;
    transition: all ${duration.status}ms ${easing.status};
    outline: none;
    &:active {
      box-shadow: ${token.activeBoxShadow};
    }
    &::after,
    ${StyledLoadingIconWrapper} {
      position: absolute;
      top: 0;
      left: 0;
      width: ${token.height[size!] - 4}px;
      height: ${token.height[size!] - 4}px;
      background-color: ${token.circleBackground};
      border-radius: 50%;
      transition: all ${duration.status}ms ${easing.status};
      content: ' ';
      cursor: ${disabled ? 'not-allowed' : 'pointer'};
      ${checked &&
        css`
          left: 100%;
          transform: translateX(-100%);
        `}
    }

    ${StyledLoadingIconWrapper} {
      text-align: center;
      font-size: ${iconFontSize}px;
      ${iconFontSize < 12 &&
        css`
          /** 小于 12 像素的文字居中问题 */
          svg {
            vertical-align: unset;
          }
        `}
    }

    ${StyledSwitchChildrenWrapper} {
      margin-left: ${checked
        ? token.childrenMarginLeft.checked
        : token.childrenMarginLeft.unChecked}px;
      margin-right: ${checked
        ? token.childrenMarginRight.checked
        : token.childrenMarginRight.unChecked}px;
      color: ${colorsPattern.icon.inverse};
      font-size: ${token.fontSize[size!]}px;
    }
  `;
};

export const StyledSwitch = styled.div`
  ${switchCss};
`;
