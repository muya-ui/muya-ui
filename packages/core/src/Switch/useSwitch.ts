import { RefObject, useRef, useState, useCallback } from 'react';

import { ISwitchProps } from './types';

export default function useSwitch(props: ISwitchProps, nodeRef: RefObject<HTMLDivElement>) {
  const {
    disabled,
    checked,
    defaultChecked,
    loading,
    onChange,
    onClick: onClickProp,
    onMouseUp: onMouseUpProp,
  } = props;
  const { current: isControlled } = useRef('checked' in props);
  const [checkedState, setCheckedState] = useState(checked || defaultChecked || false);

  const setChecked = useCallback(
    (checked: boolean, e: React.MouseEvent | React.KeyboardEvent) => {
      if (disabled) {
        return;
      }
      if (!isControlled) {
        setCheckedState(checked);
      }

      if (onChange) {
        onChange(checked, e);
      }
    },
    [disabled, isControlled, onChange],
  );

  const onKeyDown = useCallback(
    (e: React.KeyboardEvent) => {
      if (e.keyCode === 37) {
        setChecked(false, e);
        /* istanbul ignore else */
      } else if (e.keyCode === 39) {
        setChecked(true, e);
      }
    },
    [setChecked],
  );

  const onClick = useCallback(
    (e: React.MouseEvent) => {
      if (disabled || loading) {
        e.stopPropagation();
        return;
      }
      const newChecked = !checkedState;
      setChecked(newChecked, e);
      /* istanbul ignore else */
      if (onClickProp) {
        onClickProp(newChecked, e);
      }
    },
    [checkedState, disabled, loading, onClickProp, setChecked],
  );

  const blur = useCallback(() => {
    /* istanbul ignore next */
    if (nodeRef.current) {
      nodeRef.current.blur();
    }
  }, [nodeRef]);

  // Handle auto focus when click switch in Chrome
  const onMouseUp = useCallback(
    (e: React.MouseEvent) => {
      blur();
      /* istanbul ignore else */
      if (onMouseUpProp) {
        onMouseUpProp(e);
      }
    },
    [blur, onMouseUpProp],
  );

  const focus = useCallback(() => {
    /* istanbul ignore else */
    if (nodeRef.current) {
      nodeRef.current.focus();
    }
  }, [nodeRef]);

  return {
    checkedState,
    setCheckedState,
    onKeyDown,
    onClick,
    onMouseUp,
    focus,
    blur,
  };
}
