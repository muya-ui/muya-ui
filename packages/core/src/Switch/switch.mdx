---
name: Switch 开关
menu: 组件
route: /components/switch
autoDemos: true
parent: 数据录入
---

import { DocSwitchStyle } from './demos/styled';

<DocSwitchStyle />

# Switch 开关

表示两种相互对立的状态间的切换开关选择器。

## 用法

需要表示开关状态/两种状态之间的切换时使用；和 `checkbox` 的区别是，切换 `switch` 会直接触发状态改变，而 `checkbox` 一般用于状态标记，需要和提交操作配合。

{Demos}

### API

#### ISwitchProps

<Props propsName="ISwitchProps" />

### Styles

#### ISwitchStyleKeys

<Styles typeName="ISwitchStyleKeys" componentName="switch" />
