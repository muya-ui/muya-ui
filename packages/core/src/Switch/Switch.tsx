import React, { Ref, useRef, useMemo } from 'react';

import { useEffectOnce, useForkRef } from '@muya-ui/utils';

import Spin from '../Spin';
import useStyles from '../utils/useStyles';
import useTheme from '../utils/useTheme';
import { StyledLoadingIconWrapper, StyledSwitch, StyledSwitchChildrenWrapper } from './styled';
import { ISwitchProps, ISwitchStyleKeys } from './types';
import { ICustomStylePropMap } from '../types';
import useSwitch from './useSwitch';

const Switch = React.forwardRef((props: ISwitchProps, ref: Ref<HTMLDivElement>) => {
  const {
    autoFocus,
    disabled,
    checked,
    defaultChecked,
    checkedChildren,
    unCheckedChildren,
    loading,
    loadingIcon,
    onChange,
    styles,
    size = 'm',
    ...restProps
  } = props;
  const theme = useTheme();
  const {
    components: { Switch: token },
  } = theme;
  const defaultStyles = useMemo<ICustomStylePropMap<ISwitchStyleKeys>>(
    () => ({
      loadingWrapper: '',
      childrenWrapper: '',
    }),
    [],
  );
  const innerStyles = useStyles('switch', defaultStyles, styles);
  const nodeRef = useRef<HTMLDivElement>(null);
  const handleRef = useForkRef(ref, nodeRef);
  const { checkedState, setCheckedState, onKeyDown, onClick, onMouseUp, focus } = useSwitch(
    props,
    nodeRef,
  );

  useEffectOnce(() => {
    if (autoFocus && !disabled) {
      focus();
    }
  });

  if ('checked' in props && checkedState !== checked) {
    setCheckedState(checked!);
  }

  const loadingNode = useMemo(
    () =>
      loadingIcon || (
        <Spin color={checkedState ? token.background.checked : token.background.unChecked} />
      ),
    [checkedState, loadingIcon, token.background.checked, token.background.unChecked],
  );

  return (
    <StyledSwitch
      {...restProps}
      ref={handleRef}
      role="switch"
      theme={theme}
      size={size}
      checked={checkedState}
      disabled={disabled || loading}
      aria-checked={checked}
      onClick={onClick}
      onMouseUp={onMouseUp}
      onKeyDown={onKeyDown}
    >
      {loading && (
        <StyledLoadingIconWrapper {...innerStyles.loadingWrapper}>
          {loadingNode}
        </StyledLoadingIconWrapper>
      )}
      <StyledSwitchChildrenWrapper {...innerStyles.childrenWrapper}>
        {checkedState ? checkedChildren : unCheckedChildren}
      </StyledSwitchChildrenWrapper>
    </StyledSwitch>
  );
});

Switch.defaultProps = {
  checkedChildren: null,
  unCheckedChildren: null,
  defaultChecked: false,
  loading: false,
  tabIndex: 0,
};

export default React.memo(Switch);
