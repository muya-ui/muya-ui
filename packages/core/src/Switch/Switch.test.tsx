import React, { createRef } from 'react';
import renderer from 'react-test-renderer';

import { render } from '@testing-library/react';

import Switch from './index';

test('测试普通渲染', () => {
  const tree = renderer
    .create(
      <>
        <Switch defaultChecked />
        <Switch defaultChecked disabled />
        <Switch checkedChildren="开" unCheckedChildren="关" />
        <Switch defaultChecked loading />
      </>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('测试 effect', () => {
  const nodeRef = createRef();
  const { rerender } = render(<Switch ref={nodeRef} checked autoFocus />);
  rerender(<Switch ref={nodeRef} checked={false} autoFocus />);
});
