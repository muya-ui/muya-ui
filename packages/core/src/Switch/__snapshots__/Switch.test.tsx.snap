// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试普通渲染 1`] = `
Array [
  .c2 {
  display: block;
}

.c0 {
  position: relative;
  display: inline-block;
  box-sizing: border-box;
  height: 20px;
  line-height: 16px;
  min-width: 42px;
  background-color: rgba(26,122,248,1);
  border-radius: 999rem;
  vertical-align: middle;
  border: 2px solid transparent;
  opacity: 1;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  outline: none;
}

.c0:active {
  box-shadow: none;
}

.c0::after,
.c0 .c5 {
  position: absolute;
  top: 0;
  left: 0;
  width: 16px;
  height: 16px;
  background-color: rgba(255,255,255,1);
  border-radius: 50%;
  -webkit-transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  content: ' ';
  cursor: pointer;
  left: 100%;
  -webkit-transform: translateX(-100%);
  -ms-transform: translateX(-100%);
  transform: translateX(-100%);
}

.c0 .c5 {
  text-align: center;
  font-size: 8px;
}

.c0 .c5 svg {
  vertical-align: unset;
}

.c0 .c1 {
  margin-left: 8px;
  margin-right: 24px;
  color: rgba(255,255,255,1);
  font-size: 12px;
}

.c3 .c1 {
  margin-left: 8px;
  margin-right: 24px;
  color: rgba(255,255,255,1);
  font-size: 12px;
}

.c4 .c1 {
  margin-left: 24px;
  margin-right: 8px;
  color: rgba(255,255,255,1);
  font-size: 12px;
}

<div
    checked={true}
    className="c0"
    disabled={false}
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseUp={[Function]}
    role="switch"
    size="m"
    tabIndex={0}
  >
    <div
      className="c1 c2 muya-switch-children-wrapper"
    />
  </div>,
  .c2 {
  display: block;
}

.c3 .c1 {
  margin-left: 8px;
  margin-right: 24px;
  color: rgba(255,255,255,1);
  font-size: 12px;
}

.c0 {
  position: relative;
  display: inline-block;
  box-sizing: border-box;
  height: 20px;
  line-height: 16px;
  min-width: 42px;
  background-color: rgba(26,122,248,1);
  border-radius: 999rem;
  vertical-align: middle;
  border: 2px solid transparent;
  opacity: 0.3;
  cursor: not-allowed;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  outline: none;
}

.c0:active {
  box-shadow: none;
}

.c0::after,
.c0 .c5 {
  position: absolute;
  top: 0;
  left: 0;
  width: 16px;
  height: 16px;
  background-color: rgba(255,255,255,1);
  border-radius: 50%;
  -webkit-transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  content: ' ';
  cursor: not-allowed;
  left: 100%;
  -webkit-transform: translateX(-100%);
  -ms-transform: translateX(-100%);
  transform: translateX(-100%);
}

.c0 .c5 {
  text-align: center;
  font-size: 8px;
}

.c0 .c5 svg {
  vertical-align: unset;
}

.c0 .c1 {
  margin-left: 8px;
  margin-right: 24px;
  color: rgba(255,255,255,1);
  font-size: 12px;
}

.c4 .c1 {
  margin-left: 24px;
  margin-right: 8px;
  color: rgba(255,255,255,1);
  font-size: 12px;
}

<div
    checked={true}
    className="c0"
    disabled={true}
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseUp={[Function]}
    role="switch"
    size="m"
    tabIndex={0}
  >
    <div
      className="c1 c2 muya-switch-children-wrapper"
    />
  </div>,
  .c2 {
  display: block;
}

.c3 .c1 {
  margin-left: 8px;
  margin-right: 24px;
  color: rgba(255,255,255,1);
  font-size: 12px;
}

.c4 .c1 {
  margin-left: 8px;
  margin-right: 24px;
  color: rgba(255,255,255,1);
  font-size: 12px;
}

.c0 {
  position: relative;
  display: inline-block;
  box-sizing: border-box;
  height: 20px;
  line-height: 16px;
  min-width: 42px;
  background-color: rgba(204,204,204,1);
  border-radius: 999rem;
  vertical-align: middle;
  border: 2px solid transparent;
  opacity: 1;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  outline: none;
}

.c0:active {
  box-shadow: none;
}

.c0::after,
.c0 .c5 {
  position: absolute;
  top: 0;
  left: 0;
  width: 16px;
  height: 16px;
  background-color: rgba(255,255,255,1);
  border-radius: 50%;
  -webkit-transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  content: ' ';
  cursor: pointer;
}

.c0 .c5 {
  text-align: center;
  font-size: 8px;
}

.c0 .c5 svg {
  vertical-align: unset;
}

.c0 .c1 {
  margin-left: 24px;
  margin-right: 8px;
  color: rgba(255,255,255,1);
  font-size: 12px;
}

<div
    checked={false}
    className="c0"
    disabled={false}
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseUp={[Function]}
    role="switch"
    size="m"
    tabIndex={0}
  >
    <div
      className="c1 c2 muya-switch-children-wrapper"
    >
      关
    </div>
  </div>,
  .c4 {
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  font-style: normal;
  line-height: 0;
  text-align: center;
  text-transform: none;
  vertical-align: -0.125em;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  width: 1em;
  height: 1em;
  overflow: hidden;
  fill: currentColor;
}

.c5 {
  -webkit-animation: iECmZH 0.75s linear infinite,keNNsN 1.5s linear infinite;
  animation: iECmZH 0.75s linear infinite,keNNsN 1.5s linear infinite;
}

.c3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c7 {
  display: block;
}

.c2 {
  z-index: 1;
  font-size: 12px;
  background: transparent;
}

.c8::after,
.c8 .c1 {
  position: absolute;
  top: 0;
  left: 0;
  width: 16px;
  height: 16px;
  background-color: rgba(255,255,255,1);
  border-radius: 50%;
  -webkit-transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  content: ' ';
  cursor: pointer;
  left: 100%;
  -webkit-transform: translateX(-100%);
  -ms-transform: translateX(-100%);
  transform: translateX(-100%);
}

.c8 .c1 {
  text-align: center;
  font-size: 8px;
}

.c8 .c1 svg {
  vertical-align: unset;
}

.c8 .c6 {
  margin-left: 8px;
  margin-right: 24px;
  color: rgba(255,255,255,1);
  font-size: 12px;
}

.c0 {
  position: relative;
  display: inline-block;
  box-sizing: border-box;
  height: 20px;
  line-height: 16px;
  min-width: 42px;
  background-color: rgba(26,122,248,1);
  border-radius: 999rem;
  vertical-align: middle;
  border: 2px solid transparent;
  opacity: 0.3;
  cursor: not-allowed;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  outline: none;
}

.c0:active {
  box-shadow: none;
}

.c0::after,
.c0 .c1 {
  position: absolute;
  top: 0;
  left: 0;
  width: 16px;
  height: 16px;
  background-color: rgba(255,255,255,1);
  border-radius: 50%;
  -webkit-transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  content: ' ';
  cursor: not-allowed;
  left: 100%;
  -webkit-transform: translateX(-100%);
  -ms-transform: translateX(-100%);
  transform: translateX(-100%);
}

.c0 .c1 {
  text-align: center;
  font-size: 8px;
}

.c0 .c1 svg {
  vertical-align: unset;
}

.c0 .c6 {
  margin-left: 8px;
  margin-right: 24px;
  color: rgba(255,255,255,1);
  font-size: 12px;
}

.c9::after,
.c9 .c1 {
  position: absolute;
  top: 0;
  left: 0;
  width: 16px;
  height: 16px;
  background-color: rgba(255,255,255,1);
  border-radius: 50%;
  -webkit-transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  content: ' ';
  cursor: pointer;
}

.c9 .c1 {
  text-align: center;
  font-size: 8px;
}

.c9 .c1 svg {
  vertical-align: unset;
}

.c9 .c6 {
  margin-left: 24px;
  margin-right: 8px;
  color: rgba(255,255,255,1);
  font-size: 12px;
}

<div
    checked={true}
    className="c0"
    disabled={true}
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseUp={[Function]}
    role="switch"
    size="m"
    tabIndex={0}
  >
    <span
      className="c1 c2 muya-switch-loading-wrapper"
    >
      <span
        className="c3"
        direction="column"
      >
        <span
          className=""
        >
          <svg
            aria-hidden="false"
            className="c4 c5"
            focusable="false"
            preserveAspectRatio="xMidYMid"
            viewBox="0 0 100 100"
          >
            <circle
              cx="50"
              cy="50"
              fill="none"
              r="45"
              stroke="rgba(26,122,248,1)"
              strokeLinecap="round"
              strokeWidth="10"
              transform="rotate(144 50 50)"
            />
          </svg>
        </span>
        <span />
      </span>
    </span>
    <div
      className="c6 c7 muya-switch-children-wrapper"
    />
  </div>,
]
`;
